### DISCUSSION TITLE
Does Errorprone support android gradle projects?

### DISCUSSION THREAD

**Discussion by vivian196167** [tbroyer/gradle-errorprone-plugin#91] [Q&A] [answered]
We have many android application projects which use gradle 6.5 and android gradle plugin com.android.tools.build:gradle:4.2.2. I have upgrade gradle to 6.8.1, but it cannot print errors as I add the bellos code snippet.

public String test1() {
        Set<Short> s = new HashSet<>();
        for (short i = 0; i < 100; i++) {
            s.add(i);
            s.remove(i - 1);
        }
        System.out.println(s.size());
        return "Hello World!";
    }

Does error-prone support android gradle application? at what version at least should android gradle plugin be used?

**Comment by tbroyer** [comment, marked as answer]
There used to be code in the plugin to specifically support Android projects but I eventually removed it in version 3 as I could make sense of how things were moving on in AGP and how one was supposed to configure tasks with their new model.

There's a bit more context in https://github.com/tbroyer/gradle-errorprone-plugin/releases/tag/v3.0.0

TL;DR: JavaCompile tasks are enhanced to support ErrorProne, but one has to configure the dependencies and somehow enable ErrorProne on the tasks. I have no idea how to do this though 🤷
I'm encouraging anyone to make an Android-specific plugin to complement this plugin and go that extra mile.

**Reply by tsniatowski** [reply]
Hi! I think it would be good to add a short note about the lack of Android support on the main README in the plugin repo, for example around the JDK8/JDK16 support paragraphs. As-is it's not super obvious that the support has been dropped. A closed issue is not as easy to spot, and I feel people will keep asking.

**Reply by tbroyer** [reply]
Well, the README does not talk about Android, and support for Android was dropped in a major release and properly documented in the release notes.
So,
* if you're new to the plugin, it doesn't advertise support for Android (and Android builds are a very special kind of Gradle builds that requires opt-in for many plugins, so users should already know to look for explicit support)
* if you already use the plugin and upgrade from v2 to v3, then it's up to you to read the release notes looking for breaking changes

That being said, I could add a message to the README making clear that Android builds aren't supported. Note however that the README precisely describe which tasks are processed and in which condition, and that rules out Android builds (“Note that this plugin only enables Error Prone on tasks for source sets”, and Android source sets aren't regular Gradle source sets)

To be clear, I'm refraining from saying anything about Android builds in the README because I'd have a hard time not sounding mean. Talking about Android not being supported opens the question of "why", and all I can come up with as an answer is "they don't want you/us to do it". I don't do Android development myself so I have zero incentive to dive into their super-convoluted APIs to "make it work" (don't get me wrong, there are reasons why it's so complex). I believe it **is** possible to make it work, possibly using crappy workarounds (that I personally don't want to use as they'd be brittle by definition, and causing additional maintenance burden), but it is entirely possible to do it within an additional plugin. I'd be happy to link to such a plugin if it exists though (rather than explaining why this is not built-in here).
(but honestly, just use Kotlin for Android, and/or Android Lint rather than Error Prone, unless maybe you have custom Error Prone checks)

**Reply by tsniatowski** [reply]
I think a matter-of-fact note about no Android support (and for that matter, Gradle 8 support only in v3 onwards) would be a nice gesture. It's a major change and this plugin was the one to use for quite a while after all. I know I would've spotted the readme note much sooner than release notes, because I completely did not expect such a change and I thought I simply had something set up wrong in my project.

No hard feelings, of course! And thanks for the work you put in.

So basically a short note in the readme could avoid some "well-intentioned but should have searched existing bugs" issues asking for the support. I'm afraid just about now a new Android SDK is coming that will force people onto Gradle 8, which will force them to upgrade many plugins including this one, and they may end up skipping steps and heading to the issue tracker.

**Reply by tbroyer** [reply]
> and for that matter, Gradle 8 support only in v3 onwards

This is not true: the `net.ltgt.nullaway` plugin was tested with Gradle 8.0 and 8.1 against `net.ltgt.errorprone` 2.0.2: https://github.com/tbroyer/gradle-nullaway-plugin/blob/v1.6.0/build.gradle.kts#L41 / https://github.com/tbroyer/gradle-nullaway-plugin/blob/v1.6.0/.github/workflows/gradle.yaml#L42

**Reply by tbroyer** [reply]
Anyway, I've added some words about Android projects to the README: https://github.com/tbroyer/gradle-errorprone-plugin/compare/cae500031030b0e6185c74f06a7cc0e2121d9615...8d3aea3fb1b8ee457bd92b5751df4c921020a2cf

**Reply by tsniatowski** [reply]
I stand corrected, I probably confused Gradle compatibility with AGP compatibility. Anyway, thanks for the update!
